src/iop/lut3d.c

```diff
@@ -0,0 +1,1200 @@
+          interpolation_worker(out, out, clut, level);
+//          dt_linear_sRGB_to_Lab(lout, out);
+        }
+        else if (colorspace == DT_IOP_LIN_PROPHOTORGB)
+        {
+//          dt_Lab_to_prophotorgb(in, lin);
+          interpolation_worker(in, out, clut, level);
+//          dt_prophotorgb_to_Lab(lout, out);
+        }
+        else if (colorspace == DT_IOP_LOG_FUJI) // log RGB Fuji
+        {
+//          dt_Lab_to_prophotorgb(in, lin);
+          lin_to_logFuji(in, lin);
+          interpolation_worker(lin, lout, clut, level);
+          logFuji_to_lin(lout, out);
+//          dt_prophotorgb_to_Lab(lout, out);
+        }
+        in += ch;
+        out += ch;
+      }
+    }
+  }
+  else if (!(colorspace == DT_IOP_REC709 || colorspace == DT_IOP_SRGB || colorspace == DT_IOP_LIN_REC709))
+  { // no clut
+    memcpy(obuf, ibuf, roi_in->width * roi_out->height * 4 * sizeof(float));
+  }
+  if (colorspace == DT_IOP_REC709 || colorspace == DT_IOP_SRGB || colorspace == DT_IOP_LIN_REC709)
+    dt_ioppr_transform_image_colorspace_rgb(obuf, obuf, roi_out->width, roi_out->height, dest_profile, rgb_profile, "specific to linrec2020");
+}
+
+void reload_defaults(dt_iop_module_t *self)
+{
+printf("reload_defaults\n");
+}
+
+void init(dt_iop_module_t *self)
+{
+printf("init\n");
+  self->data = NULL;
+  self->params = calloc(1, sizeof(dt_iop_lut3d_params_t));
+  self->default_params = calloc(1, sizeof(dt_iop_lut3d_params_t));
+  self->default_enabled = 0;
+//  self->priority = 710; // self order created by iop_dependencies.py, do not edit!
+  self->params_size = sizeof(dt_iop_lut3d_params_t);
+  self->gui_data = NULL;
+  dt_iop_lut3d_params_t tmp = (dt_iop_lut3d_params_t)
+    { { "" },
+    DT_IOP_SRGB,
+    DT_IOP_TETRAHEDRAL,
+    50.0f,  // grey point
+    -6.0f,  // shadows range
+    10.0f,   // dynamic range
+    1  // log2 to lin
+  };
+
+  memcpy(self->params, &tmp, sizeof(dt_iop_lut3d_params_t));
+  memcpy(self->default_params, &tmp, sizeof(dt_iop_lut3d_params_t));
+}
+
+void init_global(dt_iop_module_so_t *self)
+{
+printf("init_global\n");
+  dt_iop_lut3d_global_data_t *gd
+      = (dt_iop_lut3d_global_data_t *)malloc(sizeof(dt_iop_lut3d_global_data_t));
+  self->data = gd;
+  gd->clut = NULL;
+  gd->level = 0;
+}
+
+void cleanup(dt_iop_module_t *self)
+{
+printf("cleanup\n");
+  free(self->params);
+  self->params = NULL;
+}
+
+void cleanup_global(dt_iop_module_so_t *self)
+{
+printf("cleanup_global\n");
+  dt_iop_lut3d_global_data_t *gd = (dt_iop_lut3d_global_data_t *)self->data;
+  if (gd->clut)
+  {
+    dt_free_align(gd->clut);
+    gd->clut = NULL;
+  }
+  free(self->data);
+  self->data = NULL;
+}
+
+void commit_params(struct dt_iop_module_t *self, dt_iop_params_t *p1, dt_dev_pixelpipe_t *pipe,
+                   dt_dev_pixelpipe_iop_t *piece)
+{
+printf("commit\n");
+    dt_iop_lut3d_params_t *p = (dt_iop_lut3d_params_t *)self->params;
+    dt_iop_lut3d_global_data_t *gd = (dt_iop_lut3d_global_data_t *)self->data;

```

Comment:
This should be dt_iop_lut3d_data_t, then you copy params:

memcpy(&d->params, p, sizeof(dt_iop_lut3d_params_t));

and do the calculations on d instead of gd.